MatrixSDKTests/MXClientInformationServiceUnitTests.swift

```diff
@@ -0,0 +1,165 @@
+        credentials.deviceId = mockDeviceId
+        guard let session = MXSession(matrixRestClient: MXRestClientStub(credentials: credentials)) else {
+            XCTFail("Failed to setup test conditions")
+            return
+        }
+
+        let service = MXClientInformationService(withSession: session)
+
+        let type = service.accountDataType(for: session)
+
+        session.setAccountData(["some_key": "some_value"], forType: type) {
+
+        } failure: { _ in
+            XCTFail("Failed to setup test conditions")
+        }
+
+        service.removeDataIfNeeded(on: session)
+
+        // must be empty after removeDataIfNeeded
+        let updatedInfo = session.accountData.accountData(forEventType: type)
+        XCTAssert(updatedInfo?.isEmpty ?? true)
+
+        // remove data again when empty
+        service.removeDataIfNeeded(on: session)
+
+        // must be still empty
+        let updatedInfo2 = session.accountData.accountData(forEventType: type)
+        XCTAssert(updatedInfo2?.isEmpty ?? true)
+
+        session.close()
+    }
+
+    func testRemoveDataByDisablingFeature() {
+        //  enable the feature
+        MXSDKOptions.sharedInstance().enableNewClientInformationFeature = true
+
+        let mockDeviceId = "some_device_id"
+        let credentials = MXCredentials(homeServer: "", userId: "@userid:example.com", accessToken: "")
+        credentials.deviceId = mockDeviceId
+        guard let session = MXSession(matrixRestClient: MXRestClientStub(credentials: credentials)) else {
+            XCTFail("Failed to setup test conditions")
+            return
+        }
+
+        let service = MXClientInformationService(withSession: session)
+
+        let type = service.accountDataType(for: session)
+
+        service.updateData()
+
+        let clientInfo = session.accountData.accountData(forEventType: type)
+        XCTAssertNotNil(clientInfo)
+
+        //  disable the feature
+        MXSDKOptions.sharedInstance().enableNewClientInformationFeature = false
+
+        service.updateData()
+
+        // must be empty after updateData
+        let updatedInfo = session.accountData.accountData(forEventType: type)
+        XCTAssert(updatedInfo?.isEmpty ?? true)
+
+        session.close()
+    }
+
+    func testClientInformation() {
+        //  enable the feature
+        MXSDKOptions.sharedInstance().enableNewClientInformationFeature = true
+
+        let mockDeviceId = "some_device_id"
+        let credentials = MXCredentials(homeServer: "", userId: "@userid:example.com", accessToken: "")
+        credentials.deviceId = mockDeviceId
+        guard let session = MXSession(matrixRestClient: MXRestClientStub(credentials: credentials)) else {
+            XCTFail("Failed to setup test conditions")
+            return
+        }
+
+        let service = MXClientInformationService(withSession: session)
+        let clientInfo = service.createClientInformation()
```

Comment:
It seems you are testing internal implementation method, that shouldn't be done. If you want to test values, please see the comment about testing values

Reply:
I'm not quite sure I'm getting your concern here. Why an internal method shouldn't be tested?

Reply:
When unit testing you shouldn't care about the internal implementation of the class your are testing. You are mainly interested to see what's the unit output based on the inputs you give. Pubic interface is the one that should be tested. That allows safely to refactor all the code in those classes without worrying that by changing internal implementation you may brake something. By writing enough tests that cover public interface, you are covering internal implementation as well. That is the beauty of the unit testing :)

Reply:
That was serving a purpose here but ok, i made them private again.